pkg/hubble/parser/seven/parser.go

```diff
@@ -133,31 +140,59 @@ func (p *Parser) Decode(r *accesslog.LogRecord, decoded *flowpb.Flow) error {
 	decoded.DestinationService = destinationService
 	decoded.TrafficDirection = decodeTrafficDirection(r.ObservationPoint)
 	decoded.PolicyMatchType = 0
-	decoded.TraceContext = extractTraceContext(r)
+	decoded.TraceContext = p.getTraceContext(r)
 	decoded.Summary = p.getSummary(r, decoded)
 
 	return nil
 }
 
-func (p *Parser) computeResponseTime(r *accesslog.LogRecord, timestamp time.Time) uint64 {
+func extractRequestID(r *accesslog.LogRecord) string {
 	var requestID string
 	if r.HTTP != nil {
 		requestID = r.HTTP.Headers.Get("X-Request-Id")
 	}
+	return requestID
+}
+
+func (p *Parser) getTraceContext(r *accesslog.LogRecord) *flowpb.TraceContext {
+	requestID := extractRequestID(r)
+	switch r.Type {
+	case accesslog.TypeRequest:
+		traceContext := extractTraceContext(r)
+		if traceContext == nil {
+			break
+		}
+		p.traceContextCache.Add(requestID, traceContext)
```

Comment:
It looks like we overwrite the previous value for every request that has a trace context header but no `X-Request-Id` header (ie `requestID == ""`). This seems pretty wrong to me as we might set the wrong trace ID to the flow when fetching it from the cache below.

Reply:
Good point, though I imagine we should never see the same request with the same request ID more than once, but we can add a check here, plus we don't want to be adding useless entry to the cache.

Reply:
Found an issue my tests to, I wasn't even setting traceID!

Reply:
Fixed